src/Orleans/Configuration/ProviderConfiguration.cs

```diff
@@ -193,23 +193,39 @@ public IList<IProvider> Children
     [Serializable]
     public class ProviderCategoryConfiguration
     {
-        public string Name { get; set; }
+        public ProviderCategoryConfiguration()
+        {
+            Providers = new Dictionary<string, IProviderConfiguration>();
+        }
 
-        public IDictionary<string, IProviderConfiguration> Providers { get; set; }
+        internal ProviderCategoryConfiguration(string name) : this()
```

Comment:
I have used and I have not always been satisfied. The point was that when one reviews here, how easy it to spot if there is a problem. What should be reviewed and discussed? To me it looks a bit awkward the class can be in initial states that aren't that explicit. Does one need to check also the calling code or trust the code altered does the right thing? Especially in absence of checkers or tests (we don't have code contracts complaining) -- or at least it's difficult to see tests from here. Moreover, wouldn't it be easier to refactor code that checks abudantly? In this case performance doesn't seem to be a problem or an added check doesn't feel like bringing clutter. I would err on the side of adding it, just in case and forgo a notion adding them only on public surface. 

In general, should the decision made case-by-case basis or is there harm done by just always checking constructor and function parameters? You seem to state performance and readability as reasons to avoid them in non-public surfaces, don't count internal as public surface whereas I see it as a public surface to the assembly (as well as I think protected is a somewhat public surface) and I think explicit checking increases readability (to be clear, I usually won't check parameters once using functions inside a class that have the parameters sanitized before letting them in), I'm content with that and we seem to disagree on this matter. We just need to discuss this and establish some kind of a guideline and go with that.